src/Microsoft.Extensions.Logging.EventSource/LoggingEventSource.cs

```diff
@@ -0,0 +1,235 @@
+    public class LoggingEventSource : EventSource
+    {
+        /// <summary>
+        /// This is public from an EventSource consumer point of view, but since these defintions 
+        /// are not needed outside this class 
+        /// </summary>
+        public class Keywords
+        {
+            /// <summary>
+            /// Meta events are evnets about the LoggingEventSource itself (that is they did not come from ILogger
+            /// </summary>
+            public const EventKeywords Meta = (EventKeywords)1;
+            /// <summary>
+            /// Turns on the 'Message' event when ILogger.Log() is called.   It gives the information in a programatic (not formatted) way  
+            /// </summary>
+            public const EventKeywords Message = (EventKeywords)2;
+            /// <summary>
+            /// Turns on the 'FormatMessage' event when ILogger.Log() is called.  It gives the formatted string version of the information.  
+            /// </summary>
+            public const EventKeywords FormattedMessage = (EventKeywords)4;
+            /// <summary>
+            /// Turns on the 'MessageJson' event when ILogger.Log() is called.   It gives  JSON representation of the Arguments. 
+            /// </summary>
+            public const EventKeywords JsonMessage = (EventKeywords)8;
+        }
+
+        /// <summary>
+        ///  The one and only instance of the LoggingEventSource.  
+        /// </summary>
+        internal static readonly LoggingEventSource Instance = new LoggingEventSource();
+
+        internal static readonly LogLevel LoggingDisabled = LogLevel.None + 1;
+
+        private string _filterSpec;
+        private EventSourceLoggerProvider _loggingProviders;
+        private object _lockObj = new object();
+
+        internal EventSourceLoggerProvider CreateLoggerProvider(ILoggerFactory factory)
+        {
+            lock (_lockObj)
+            {
+                var newLoggerProvider = new EventSourceLoggerProvider(factory, this, _loggingProviders);
+                _loggingProviders = newLoggerProvider;
+
+                // If the EventSource has already been turned on.  set the filters.  
+                if (_filterSpec != null)
+                {
+                    newLoggerProvider.SetFilterSpec(_filterSpec);
+                }
+
+                return newLoggerProvider;
+            }
+        }
+
+#if !NO_EVENTSOURCE_COMPLEX_TYPE_SUPPORT
+        private LoggingEventSource() : base(EventSourceSettings.EtwSelfDescribingEventFormat) { }
+#endif
+
+        /// <summary>
+        /// FormattedMessage() is called when ILogger.Log() is called. and the FormattedMessage keyword is active
+        /// This only gives you the human reasable formatted message. 
+        /// </summary>
+        [Event(1, Keywords = Keywords.FormattedMessage, Level = EventLevel.LogAlways)]
+        internal void FormattedMessage(LogLevel Level, int FactoryID, string LoggerName, string EventId, string FormattedMessage)
```

Comment:
While we use a single ETW Event, the LogLevel is part of event payload. From practical standpoint (of indexing and querying the event data) it makes very little difference whether the event level information is expressed through ETW event level or through custom payload property (like here). So I suggest that what we have is "good enough" for now. We can certainly revisit this suggestion if we get more feedback about this.